overlord/servicestate/servicestate.go

```diff
@@ -41,38 +42,141 @@ type Instruction struct {
 
 type ServiceActionConflictError struct{ error }
 
+// changeIDForContext returns change ID for non-ephemeral context
+// or empty string if context is nil or ephemeral.
+func changeIDForContext(context *hookstate.Context) string {
+	if context != nil && !context.IsEphemeral() {
+		if task, ok := context.Task(); ok {
+			if chg := task.Change(); chg != nil {
+				return chg.ID()
+			}
+		}
+	}
+	return ""
+}
+
+// serviceControlTs creates "service-control" task for every snap derived from appInfos.
+func serviceControlTs(st *state.State, appInfos []*snap.AppInfo, inst *Instruction) (*state.TaskSet, error) {
+	servicesBySnap := make(map[string][]string)
+	for _, app := range appInfos {
+		if !app.IsService() {
+			// this function should be called with services only
+			return nil, fmt.Errorf("internal error: %s is not a service", app.Name)
+		}
+		snapName := app.Snap.InstanceName()
+		servicesBySnap[snapName] = append(servicesBySnap[snapName], app.Name)
+	}
+
+	ts := state.NewTaskSet()
+	var prev *state.Task
+	for snapName, services := range servicesBySnap {
+		var snapst snapstate.SnapState
+		if err := snapstate.Get(st, snapName, &snapst); err != nil {
+			return nil, err
+		}
+		info, err := snapst.CurrentInfo()
+		if err != nil {
+			return nil, err
+		}
+		if len(info.Services()) == 0 {
+			return nil, fmt.Errorf("snap %q does not have services", snapName)
+		}
+
+		cmd := &ServiceAction{SnapName: snapName}
+		switch {
+		case inst.Action == "start":
+			cmd.Action = "start"
+			if inst.Enable {
+				cmd.ActionModifier = "enable"
+			}
+		case inst.Action == "stop":
+			cmd.Action = "stop"
+			if inst.Disable {
+				cmd.ActionModifier = "disable"
+			}
+		case inst.Action == "restart":
+			if inst.Reload {
+				cmd.Action = "reload-or-restart"
+			} else {
+				cmd.Action = "restart"
+			}
+		default:
+			return nil, fmt.Errorf("unknown action %q", inst.Action)
+		}
+
+		var svcs []string
+		for _, svcName := range services {
+			if svcName == snapName {
+				// all services of the snap
+				svcs = nil
+				break
+			}
+			app, ok := info.Apps[svcName]
+			if !(ok && app.IsService()) {
+				return nil, fmt.Errorf(i18n.G("unknown service: %q"), svcName)
+			}
+			svcs = append(svcs, app.Name)
+		}
+		cmd.Services = svcs
+
+		var summary string
+		if len(svcs) > 0 {
+			summary = fmt.Sprintf("Run service command %q for services %q of snap %q", cmd.Action, svcs, cmd.SnapName)
+		} else {
+			summary = fmt.Sprintf("Run service command %q for services of snap %q", cmd.Action, cmd.SnapName)
+		}
+
+		task := st.NewTask("service-control", summary)
```

Comment:
Should services from separate snaps be managed in separate lanes?

Reply:
I'm not sure that is useful here?  people except the whole thing to work or not. In the case of hooks we are already inside a lane I think cc @stolowski

Reply:
Yes, I'm sorry I left this unanswered, I didn't miss it, was just pondering a lot about it. TBH I'm not sure there is one answer to this, but I'm generally leaning towards Samuele's view. As for hooks, yes, they are in lanes for multi-snap ops.

Reply:
It does make sense to use one lane for the all-or-nothing approach. Can we leave a comment for that please.

Reply:
@zyga you'll have to elaborate what you mean there, because lane exists exactly for the opposite, to make some things fail and others not in one change. I probably should read the code again.

Reply:
There are 3 usage cases ATM I think:

* from the configure hook,  in that case all the tasks a postponed but put in the same lane as the configure hook
* from other hooks,  a new change is executed synchronously, anyway all the services are from one snap
* used for one of the snap start|stop|... commands  that's the only case lanes might make sense but we would also need to know we are in that case, also it would be a change from current behavior